aws/s3/s3_service.go

```diff
@@ -0,0 +1,201 @@
+package s3
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"path"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/feature/s3/manager"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+)
+
+type BucketLists struct {
+	*s3.ListBucketsOutput
+}
+
+type FileLists struct {
+	*s3.ListObjectsV2Output
+}
+
+// Struct for storing the bucket and key for an item on s3.
+type Item struct {
+	Bucket string `json:"bucket"`
+	Key    string `json:"key"`
+}
+
+// Return a URI representation of an S3Item struct.
+func (s3Key *Item) URI() string {
+	path := path.Join(s3Key.Bucket, s3Key.Key)
+
+	s3URI := fmt.Sprintf("s3://%s", path)
+
+	return s3URI
+}
+
+// Parses an s3 URI for easy access to its different components.
+// We store our s3 keys as URIs, to include all details necessary for retrieval.
+func ParseS3URI(s3URI string) (*Item, error) {
+	u, err := url.Parse(s3URI)
+	if err != nil {
+		return nil, err
+	}
+
+	item := Item{
+		Bucket: u.Host,
+		Key:    u.Path,
+	}
+
+	return &item, nil
+}
+
+func CreateBucket(client *s3.Client, bucket string) (*string, error) {
```

Comment:
Is it better to pass the client every time?
or use composition and have method and client in a specific struct?

```
type PhilS3 struct {
	Client *s3.Client
}


func (ps3 PhilS3)CreateBucket(bucket string) (*string, error) {
	input := &s3.CreateBucketInput{
		Bucket: aws.String(bucket),
	}

	_, err := ps3.Client.CreateBucket(context.Background(), input)
	if err != nil {
		return nil, err
	}

	// Before returning the key, create an s3 key (URI)
	item := Item{
		Bucket: bucket,
	}
	s3URI := item.URI()

	return &s3URI, nil
} 


```